File: JavaScript/hangman.js
const readlineSync = require('readline-sync');
const fs = require('fs');

let totalScore = 0;
const ScoreFileName = 'scores.txt';

function main() {
    while (true) {
        console.clear();
        console.log("Welcome to Hangman!");
        console.log("1. Start Game");
        console.log("2. View Score");
        console.log("3. View All Scores");
        console.log("4. Exit");
        
        const choice = prompt("\nEnter your choice: ").trim();

        switch (choice) {
            case "1":
                startGame();
                break;
            case "2":
                viewScore();
                break;
            case "3":
                viewAllScores();
                break;
            case "4":
                console.log("Thank you for playing Hangman! Goodbye!");
                return;
            default:
                console.log("Invalid choice. Please select 1, 2, 3, or 4.");
                prompt("Press Enter to continue...");
                break;
        }
    }
}

function startGame() {
    const categories = {
        1: { name: "Animals", words: ["lion", "elephant", "giraffe", "dolphin", "penguin", "tiger", "kangaroo", "panda", "eagle"] },
        2: { name: "Countries", words: ["germany", "iceland", "egypt", "france", "brazil", "japan", "canada", "italy", "australia"] },
        3: { name: "Food", words: ["pancakes", "cookies", "pizza", "burger", "sushi", "tacos", "pasta", "salad", "steak", "burek"] },
        4: { name: "Sports", words: ["football", "tennis", "cricket", "basketball", "baseball", "swimming", "volleyball", "hockey"] }
    };

    console.clear();
    console.log("Choose a category:");
    for (const key in categories) {
        console.log(`${key}. ${categories[key].name}`);
    }

    let chosenCategoryKey;
    while (true) {
        chosenCategoryKey = parseInt(prompt("\nEnter the number of your choice: ").trim());
        if (!isNaN(chosenCategoryKey) && categories[chosenCategoryKey]) {
            break;
        } else {
            console.log("Invalid choice. Please enter a valid number.");
        }
    }

    const category = categories[chosenCategoryKey];
    const wordToGuess = category.words[Math.floor(Math.random() * category.words.length)];
    let guessedWord = Array(wordToGuess.length).fill('_');
    let guessedLetters = new Set();
    const maxAttempts = 6;
    let attempts = 0;

    console.log(`\nYou selected the category: ${category.name}`);
    console.log("Let's start the game!");

    while (attempts < maxAttempts && guessedWord.join('') !== wordToGuess) {
        console.clear();
        displayHangman(attempts);
        console.log(`\nCategory: ${category.name}`);
        console.log(`\nWord: ${guessedWord.join(' ')}`);
        console.log(`Attempts left: ${maxAttempts - attempts}`);
        console.log(`Guessed letters: ${Array.from(guessedLetters).join(', ')}`);
        
        const input = prompt("Guess a letter: ").toLowerCase();
        if (!input || input.length !== 1 || !/[a-z]/.test(input)) {
            console.log("Please enter a single valid letter.");
            prompt("Press Enter to continue...");
            continue;
        }

        if (guessedLetters.has(input)) {
            console.log("You've already guessed that letter.");
            prompt("Press Enter to continue...");
            continue;
        }

        guessedLetters.add(input);

        if (wordToGuess.includes(input)) {
            for (let i = 0; i < wordToGuess.length; i++) {
                if (wordToGuess[i] === input) {
                    guessedWord[i] = input;
                }
            }
        } else {
            attempts++;
        }
    }

    console.clear();
    displayHangman(attempts);
    if (guessedWord.join('') === wordToGuess) {
        const roundScore = Math.max(10 - attempts, 0);
        console.log(`\nCongratulations! You've guessed the word: ${wordToGuess}`);
        console.log(`You earned ${roundScore} points this round.`);
        totalScore += roundScore;

        if (prompt("\nWould you like to save your score? (y/n): ").toLowerCase() === 'y') {
            saveScore(roundScore);
            console.log("Score saved successfully!");
        }
    } else {
        console.log(`\nGame over! The word was: ${wordToGuess}`);
    }
    prompt("\nPress Enter to return to the main menu.");
}

function viewScore() {
    console.clear();
    console.log(`Your total score is: ${totalScore}`);
    prompt("\nPress Enter to return to the main menu.");
}

function viewAllScores() {
    console.clear();
    console.log("All Scores:");
    if (fs.existsSync(ScoreFileName)) {
        const allScores = fs.readFileSync(ScoreFileName, 'utf8').split('\n');
        allScores.forEach(line => console.log(line));
    } else {
        console.log("No scores recorded yet.");
    }
    prompt("\nPress Enter to return to the main menu.");
}

function saveScore(roundScore) {
    const playerName = prompt("Enter your name: ").trim();
    if (playerName) {
        const scoreEntry = `${playerName}: ${roundScore} points`;
        fs.appendFileSync(ScoreFileName, scoreEntry + '\n');
    } else {
        console.log("Score not saved. Name cannot be empty.");
    }
}

function displayHangman(attempts) {
    const hangmanStages = [
        `
     +---+
     |   |
         |
         |
         |
         |
  =========`,
        `
     +---+
     |   |
     O   |
         |
         |
         |
  =========`,
        `
     +---+
     |   |
     O   |
     |   |
         |
         |
  =========`,
        `
     +---+
     |   |
     O   |
    /|   |
         |
         |
  =========`,
        `
     +---+
     |   |
     O   |
    /|\\  |
         |
         |
  =========`,
        `
     +---+
     |   |
     O   |
    /|\\  |
    /    |
         |
  =========`,
        `
     +---+
     |   |
     O   |
    /|\\  |
    / \\  |
         |
  =========`
    ];
    console.log(hangmanStages[attempts]);
}

function prompt(query) {
    return readlineSync.question(query);
}

main();

File: Python/hangman.py
import time
from random import randint
import os
import copy

HANGMANPICS = ['''
  +---+
  |   |
      |
      |
      |
      |
=========''', '''
  +---+
  |   |
  O   |
      |
      |
      |
=========''', '''
  +---+
  |   |
  O   |
  |   |
      |
      |
=========''', '''
  +---+
  |   |
  O   |
 /|   |
      |
      |
=========''', '''
  +---+
  |   |
  O   |
 /|\  |
      |
      |
=========''', '''
  +---+
  |   |
  O   |
 /|\  |
 /    |
      |
=========''', '''
  +---+
  |   |
  O   |
 /|\  |
 / \  |
      |
=========''']

categories=['Animals', 'Countries', 'Food', 'Sports']
words=[['LION','ELEPHANT','TIGER','GIRAFFE','DOLPHIN','PENGUIN','KOALA','KANGAROO','ZEBRA','CHEETAH','GORILLA','HIPPOPTAMUS', 'CROCODILE', 'CHIMPANZEE', 'LIZARD']
,['CANADA', 'JAPAN', 'BRAZIL', 'AUSTRIA', 'FRANCE', 'MEXICO', 'GERMANY', 'PAKISTAN', 'ITALY', 'CHINA', 'RUSSIA', 'SPAIN', 'EGYPT', 'CROATIA', 'SERBIA']
,['PIZZA', 'SUSHI','BURGER', 'PANCAKE', 'SPAGHETTI', 'TACOS', 'BURRITO', 'SALAD', 'BUREK', 'CHICKEN', 'STEAK', 'SOUP', 'PASTA', 'SANDWICH','RAMEN']
,['FOOTBALL','BASKETBALL','TENNIS','SWIMMING', 'VOLLEYBALL', 'BASEBALL', 'GOLF', 'RUGBY', 'KICKBOX', 'GYMNASTICS', 'MMA', 'SAILING', 'SKIING', 'JUDO','ARCHERY']]

score=0
file_name="highscores.txt"

def Score(file_name):
  with open(file_name, 'r') as file:
    print(file.read())
  time.sleep(5)

def SaveScore(file_name, name, score):
    if not os.path.exists(file_name):
        with open(file_name, 'w') as file:
            file.write("=== High Scores ===\n") 
    
    scores = []
    with open(file_name, 'r') as file:
        lines = file.readlines()
        for line in lines[1:]:  
            line = line.strip()
            if line:
                entry_name, entry_score = line.rsplit(' - ', 1)
                scores.append((entry_name, int(entry_score)))
    
    scores.append((name, score))
    
    scores.sort(key=lambda x: x[1], reverse=True)
    
    with open(file_name, 'w') as file:
        file.write("=== High Scores ===\n") 
        for idx, (entry_name, entry_score) in enumerate(scores, start=1):
            file.write(f"{idx}. {entry_name} - {entry_score}\n")

def AskScore():
  answer=input("Do you want to save the score?(Y/N):")
  if(answer=='Y' or answer=='y'):
    name=input("Write your name:")
    SaveScore(file_name,name,score)
  elif(answer=='N' or answer=='n'):
    pass
  else:
    print("You enter the wrong letter!")

def GuessWord(word):
  guessedLetters=[]
  guess=" "
  IsGuessed=False
  numberOfGuess=0
  global score
  for letter in word:
    guess=guess+"_ "
  while not(IsGuessed):
    print(f"{HANGMANPICS[numberOfGuess]}")
    print(f"{guess}")
    print(f"Attempts left: {6-numberOfGuess}")
    print(f"Guessed letters: {guessedLetters}")
    letter=input("Please make the guess:")
    while not(letter.isalpha()):
      letter=input("Invalid input! Please enter the valid input:")
    guessedLetters.append(letter)
    if(word.find(letter.upper())!=-1):
      index=word.find(letter.upper())
      while (word[index:].find(letter.upper())!=-1):
        if(word[index]==letter.upper()):
          newIndex=2*index+1
          guess=guess[:newIndex]+letter.upper()+guess[newIndex+1:]
        index=index+1
    else:
      numberOfGuess=numberOfGuess+1
    if(guess.find('_')==-1):
      IsGuessed=True
      print("|========================================|")
      print(f"You can guess your word.\nIt is {guess}")
      score+=(10-numberOfGuess)
      time.sleep(2)
      return True

    if(numberOfGuess>5):
      IsGuessed=True
      print(f"{HANGMANPICS[numberOfGuess]}")
      print("You are the loser!")
      print(f"The correct word is:{word}")
      return False

def Process(number):
  num=int(number)
  array=copyWords[num-1]
  copyWords.pop(num-1)
  while True:
    if(len(array)==0):
      print("You guess all words")
      return True
    else:
      randomIndex=randint(0,len(array)-1)
      randomWord=array[randomIndex]
      array.pop(randomIndex)
      IsGoing=GuessWord(randomWord)
      if IsGoing==False:
        print(f"Your score:{score}")
        time.sleep(2)
        AskScore()
        return False
  
def StartGame():
  wrongWord=True
  while len(copyCategories)!=0 and wrongWord==True:
    print("Choose the category for guess:")
    for i in range(0,len(copyCategories)):
      print(f"{i+1}.{copyCategories[i]}")
    add=input("Write the number to select the category:")
    copyCategories.pop(int(add)-1)
    wrongWord=Process(add)

def game():
  os.system('cls')
  global score
  global copyCategories
  global copyWords
  ad='0'
  while ad!='3':
    print("1.Start Game")
    print("2.Score")
    print("3.Exit")
    ad=input("Write the number to enter:")
    while not(ad.isdigit()):
      ad=input("Invalid input! Please enter the valid input:")
    if ad=='1':
      score=0
      copyCategories=copy.copy(categories)
      copyWords=copy.deepcopy(words)
      StartGame()
    elif ad=='2':
      Score(file_name)
    elif ad=='3':
      exit()
    else:
      print("You typed a wrong number please, try again!")
    print("----------Welcome back to the lobby!----------")

print("----------Hello and welcome to the Hangman!----------")
game()